sky/sdk/lib/widgets/block_viewport.dart

```diff
@@ -183,22 +261,47 @@ class BlockViewport extends RenderObjectWrapper {
   }
 
   void layout(BoxConstraints constraints) {
-    if (!_dirty)
+    if (!_dirty && layoutState.isValid)
       return;
     _dirty = false;
 
     Map<_Key, Widget> newChildren = new Map<_Key, Widget>();
     Map<int, Widget> builtChildren = new Map<int, Widget>();
 
+    final List<double> offsets = layoutState._childOffsets;
     final double height = root.size.height;
     final double endOffset = startOffset + height;
     BoxConstraints innerConstraints = new BoxConstraints.tightFor(width: constraints.constrainWidth());
 
+    // Before doing the actual layout, fix the offsets for the widgets
+    // whose size or type has changed.
+    if (!layoutState.isValid && offsets.length > 0) {
+      List<int> invalidIndices = layoutState._invalidIndices.toList();
+      invalidIndices.sort();
+      if (invalidIndices.last < offsets.length - 1)
+        invalidIndices.add(offsets.length - 1);
+      for (int i = 0; i < invalidIndices.length - 1; i++) {
```

Comment:
I agree that ++i is ugly, and I'll admit to being a member of the K&R i++ cargo cult. And I LIKE i += 1. Presently we've got about a dozen of "for (int i = 0; i < foo; i++)" and roughly another dozen with ++i. I will be the first i += 1 :-).